internal/types/cloud_init.go

```diff
@@ -0,0 +1,204 @@
+package types
+
+import (
+	"bytes"
+	_ "embed"
+	"encoding/base64"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+
+	"github.com/kdomanski/iso9660/util"
+
+	"github.com/cockroachdb/errors"
+)
+
+var (
+	//go:embed templates/user-data.yaml
+	userData string
+	//go:embed templates/meta-data.yaml
+	metaData string
+	//go:embed templates/network-config.yaml
+	networkData string
+
+	userDataTpl *template.Template
+	metaDataTpl *template.Template
+	networkTpl  *template.Template
+)
+
+type CloudInitGateway struct {
+	IP     string `json:"ip"`
+	OnLink bool   `json:"on_link"`
+}
+
+type CloudInitConfig struct {
+	// use remote server to fetch cloud-init config
+	URL string `json:"url"`
+
+	// use local iso to fetch cloud-init config
+	Username   string            `json:"username"`
+	Password   string            `json:"password"`
+	SSHPubKey  string            `json:"ssh_pub_key"`
+	Hostname   string            `json:"hostname"`
+	InstanceID string            `json:"instance_id"`
+	Files      map[string][]byte `json:"files"`
+	Commands   []string          `json:"commands"`
+
+	MAC       string           `json:"-"`
+	CIDR      string           `json:"-"`
+	MTU       int              `json:"-"`
+	IFName    string           `json:"-"`
+	DefaultGW CloudInitGateway `json:"-"`
+}
+
+func initTpls() (err error) {
```

Comment:
how about `func init()` then we can remove all the `if xxx == nil` from the func

Reply:
or `sync.Once(initTpls)`

Reply:
I don't like `init`,  I always run initialization code explicitly, `sync.Once` seems good